### PR TITLE
Improved Elixir optlib parser

### PR DESCRIPTION
**Pull request by dreamtigers**
* Took the branch `elixir-optlib` and rebased it to a most recent master.
* Added some spaces in `optlib2c` for a better looking C code.
* Improved the Elixir parser `elixir.ctags` ([sent a pull request to the upstream](https://github.com/mmorearty/elixir-ctags/pull/8) but made the same improvements here, in fewer commits for conciseness)
* Added a more exhaustive test case for the new parser, as requested in #1806.

### CODE REVIEW

File: makefiles/optlib2c_input.mak [hunk 1]
```diff
@@ -3,7 +3,8 @@ OPTLIB2C_INPUT = \
 	optlib/RSpec.ctags			\
 	optlib/cmake.ctags			\
 	optlib/ctags-optlib.ctags		\
-	optlib/elm.ctags			\
+	optlib/elixir.ctags		\
+	optlib/elm.ctags		\
```
Comment (masatake) [inline makefiles/optlib2c_input.mak]: You should not touch elm.ctags.

File: misc/optlib2c [hunk 2]
```diff
@@ -826,22 +826,22 @@ EOF
     }
     if (@{$opts->{'kinddefs'}}) {
 	print <<EOF;
-	def->kindTable = $opts->{'Clangdef'}KindTable;
-	def->kindCount = ARRAY_SIZE($opts->{'Clangdef'}KindTable);
+	def->kindTable     = $opts->{'Clangdef'}KindTable;
+	def->kindCount     = ARRAY_SIZE($opts->{'Clangdef'}KindTable);
```
Comment (masatake) [inline misc/optlib2c]: These changes about whitespace have impacts on ALL generated code under optlib/ directory.
Before adding Elixir related change, you should make a pull request for this change. I will merge it first.
The pull request should include all changes under optlib/*.c, too.

File: Units/simple-elixir.d/input.ex [hunk 3]
```diff
@@ -1,18 +1,190 @@
...
+# TODO: I don't know how to identificate/differenciate an exception since they
+# don't _actually_ have name. Instead are defined like a struct.
+#
+# e exceptions      (defexception ...)
+#
+# defmodule MyAppError do
+#   defexception [:message]
+#
+#   @impl true
+#   def exception(value) do
```
Comment (masatake) [inline Units/simple-elixir.d/input.ex]: An anonymous tag is needed here.  ctags main part should provide such a feature. 
This is my job.

File: Units/simple-elixir.d/input.ex [hunk 4]
```diff
@@ -1,18 +1,190 @@
...
+# by Ivan Gonzalez Polanco <ivan14polanco@gmail.com>
 #
-# Taken from https://elixir-lang.org/
+# This code doesn't compile, since it's supposed to test de ctags elixir
+# parser, not the elixir elixir parser.
+
+#
+# f functions (def ...)
+#
+defmodule MyString do
```
Comment (masatake) [inline Units/simple-elixir.d/input.ex]: Just a question. defmodule can be nested?
If not, you can build a scope. I will make more advises later.

**Comment by masatake** [comment]
You can run the test with "make units LANGUAGES=Elixir" locally.

File: Units/simple-elixir.d/input.ex (hunk 3 above)
Comment (masatake) [reply Units/simple-elixir.d/input.ex]: See #2025.

File: Units/simple-elixir.d/input.ex (hunk 4 above)
Comment (masatake) [reply Units/simple-elixir.d/input.ex]: See a test case in #2025. It is an example of how scope flags work.

**Comment by masatake** [comment]
Could you consider a ".b" test case for the TODO items?
See .b test in "Gathering test cases for known bugs" in http://docs.ctags.io/en/latest/units.html.

Some valuable studies about changing the regex engine used in u-ctags are in #1861.

File: makefiles/optlib2c_input.mak (hunk 1 above)
Comment (dreamtigers) [reply makefiles/optlib2c_input.mak]: Yes, this was a conflict when rebasing 09861b6 to `master`.
The tabs in the line `optlib/elm.ctags` were corrected in 806f296, but that commit happened _after_ 09861b6, which triggered the conflict.

I should have seen and corrected the tabs, tho.

Comment (dreamtigers) [reply makefiles/optlib2c_input.mak]: (trying out Github's suggestions feature, sorry if it's messy)
```suggestion
	optlib/elixir.ctags		\
	optlib/elm.ctags		\
```

File: Units/simple-elixir.d/input.ex (hunk 4 above)
Comment (dreamtigers) [reply Units/simple-elixir.d/input.ex]: Yes, defmodule can be nested.

File: misc/optlib2c (hunk 2 above)
Comment (dreamtigers) [reply misc/optlib2c]: Pull request done. There wasn't any other change under optlib/*.c, except for `optlib/elixir.c` but I think that change remains here (please correct me if I'm wrong).

How do I fix the fact that pull request #2026 and commit ffd1e98e contain the same changes? Sorry for the mess!

**Comment by dreamtigers** [comment]
I renamed the `simple-elixir.d` directory to `parser-elixir.r`. Divided the test case in smaller cases, so it was worth it to rename the directory (as specified in `docs/units.rst`, in the section 'Categories'). And fixed one of the TODO's where the parser wasn't generating the right tags for the word logical operators (and,or,not, etc...).

File: docs/news.rst [hunk 5]
```diff
@@ -52,6 +52,7 @@ The following parsers have been added:
 * Diff
 * DTD
 * DTS
+* Elixir *optlib*
```
Comment (masatake) [inline docs/news.rst]: Excellent.

**Comment by masatake** [comment]
Thank you. I will take over this. I will merge this after squashing some commits to keep change history simple. I'm in busy now. So, it will take few days.

**Comment by masatake** [comment]
Merged after rearranging the commits posted this PR. See #2034.
Thank you.

**Comment by dreamtigers** [comment]
Thanks a lot for the support and improving the parser! Much appreciated!

**Pull request closed without merging**